### PR TITLE
[Kernel] Add `w4a16` support for `compressed_tensors` models

### PR DESCRIPTION
**Pull request by dsikka**
# Summary
- Adds an additional `CompressedTensorsScheme` to supoprt `w4a16` models, `CompressedTensorsW4A16`
- Add support for `w4a16` models quantized through [sparseml](https://github.com/neuralmagic/sparseml) and saved through [compressed-tensors](https://github.com/neuralmagic/compressed-tensors)
- Expands `CompressedTensorsConfig`  to use `QuantizationArgs`, `QuantizationStrategy` and `find_first_name_or_class_match` to help match the appropriate scheme to each layer. This is done by adding [compressed-tensors](https://pypi.org/project/compressed-tensors/) as a requirement
- Leverages the kernels (`gptq_marlin_gemm`) used for `gptq_marlin`

From Neural Magic, Co-authored by @alexm-neuralmagic

### CODE REVIEW

File: requirements-common.txt [hunk 1]
```diff
@@ -19,3 +19,4 @@ lm-format-enforcer == 0.10.1
 outlines == 0.0.34 # Requires torch >= 2.1.0
 typing_extensions
 filelock >= 3.10.4 # filelock starts to support `mode` argument from 3.10.4
+compressed-tensors == 0.3.3
```
Comment (robertgshaw2-neuralmagic) [inline requirements-common.txt:22]: this should be in requirements-cuda?

File: vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w4a16.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+from vllm.model_executor.layers.quantization.compressed_tensors.schemes import (
+    CompressedTensorsScheme)
+from vllm.model_executor.layers.quantization.gptq_marlin import (
+    GPTQMarlinState, marlin_permute_scales)
+from vllm.model_executor.utils import set_weight_attrs
+
+__all__ = ["CompressedTensorsW4A16"]
+
+
+class CompressedTensorsW4A16(CompressedTensorsScheme):
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w4a16.py:16]: Does this do only W4 or also W8?

Comment (dsikka) [reply vllm/model_executor/layers/quantization/compressed_tensors/schemes/compressed_tensors_w4a16.py:16]: Will be handled in a follow-up PR

File: requirements-cuda.txt [hunk 3]
```diff
@@ -7,3 +7,4 @@ nvidia-ml-py # for pynvml package
 torch == 2.3.0
 xformers == 0.0.26.post1  # Requires PyTorch 2.3.0
 vllm-flash-attn == 2.5.8.post2  # Requires PyTorch 2.3.0
+compressed-tensors == 0.3.3
```
Comment (robertgshaw2-neuralmagic) [inline requirements-cuda.txt:10]: I don't think we can have this as a dependency until we remove the transformers dep from compressed tensors

**Pull request closed without merging**